### DISCUSSION TITLE
Why SingleFileZ is a standalone project?

### DISCUSSION THREAD

**Discussion by scruel** [gildas-lormeau/SingleFileZ#162] [Q&A] [answered]
Why SingleFileZ is a standalone project?
I think this should be an option to SingleFile, because it just provides a save in which type option(zip or not zip html file)?

**Comment by gildas-lormeau** [comment]
SingleFileZ is a standalone project for historical, practical and "marketing" reasons. 

When I "discovered" that an extension like SingleFile could produce HTML/ZIP files, I was tempted to add this feature in SingleFile itself, as an option. However, I quickly noticed it was a bit risky because the changes to apply in the code to support the legacy format and this new format would have been too large. So, I decided to fork SingleFile in order to deliver faster this feature without taking too much risks. I also thought that some SingleFile users would not like this option or would be confused because it is necessary to have JavaScript enabled to read a web page. Moreover, there are still some functional differences today, the annotation editor is still missing from SingleFileZ and SingleFile can't save password-protected pages for example.

I was also hoping, perhaps wrongly, that issuing a separate extension would help to highlight and publicize this alternative technique for saving pages. I was quite curious to see the feedback from the "technical" community about the innovation brought by SingleFileZ. On that point, it's a bit of a failure. I didn't get much feedback. Maybe I overestimated my ability to promote SingleFileZ to them.

Finally, (rightly?) I was never confident that self-extracting HTML files would work on Chrome in the long run (when opened from the filesystem). At first, the extension was only available for Firefox and I didn't plan to make it compatible with Chrome. This also motivated my decision to create a separate extension.

**Comment by scruel** [comment]
I've been using your extension SingleFile for almost three years, cause the workflow I recently followed on my writing progress as https://github.com/gildas-lormeau/SingleFile/issues/819 mentioned, the plugin of browser can not elegantly fit in. 
To fit in my new workflow, write an automatic script which invoke your plugin is one way to achieve my goal, then I noticed that you have finished a CLI tool (nice work!) which I needed.
But CLI with selenium(should I try other WebDriver?) seems not stable enough(https://github.com/gildas-lormeau/single-file-cli/issues/46), so I want to find another way. I read the README again, then SingleFileZ comes up to me. (finally, I choose to keep use the plugin via browser)

I didn't notice any info about SingleFileZ before, like the other ones, they will not know about it until they completely read the README of SingleFile on GitHub.
So if you want to let more users know about this tech, I think you should consider add some description in SingleFile settings page, or add it as an experiment option which users can easily aware this alternative technique for saving pages(IDK will mozilla or chrome allow this behaviour)

I know that the different part of this forked project is not just about saving the content in different ways, for example, convert the image info to base64 or binary, but the process of processing should be able to abstract, it should not have any risks. From my opinion, the plan "deliver the new format tech faster" is accomplished, when saving in new format, you can check if JavaScript enabled or not, if not, it can ask "should save it in old format", for some website which I disabled JavaScript, this will be helpful. I hope that this two project can be combined into to one, after this you will not need to copy the code from SingleFile to SingleFileZ anymore.

For now, I still keep using SingleFile for saving some websites I disabled the JavaScript, but I prefer to use SingleFileZ, because it can save disk space, and when I want to edit the saved HTML files, I do not have to face the Base64 encoded images in the source code, and can view the images(or other resources) more convenient.

I like Firefox for daily use, we can not force Chrome and other browsers team to support this new format(when opened from the file system), but display a notice when switching option to new format should be enough.

**Comment by gildas-lormeau** [comment]
If I remember correctly, promoting other extensions on the stores is not allowed. I'll check it out. However, it doesn't really matter that the discovery of the extension is progressive. That's what allowed me to not feel pressured during the development of SingleFileZ.

Thank you for your feedback, I completely understand your point of view. I also agree with you that a single extension would certainly make things easier for users. Note that it will require some work to merge back the changes into SingleFile. I have made efforts to ensure that the code bases diverge as little as possible but I think there are several tens of hours of work and testing to achieve this goal. It will also require a lot of rigor to minimize possible regressions.

**Comment by maliayas** [comment]
SingleFileZ is a great innovation. As an old maff user, I discovered it by coincidence. Apparently some marketing is needed. Maybe it should appear in places like below so that _relevant_ people become aware of it:

- https://support.mozilla.org/en-US/questions/1178293
- https://alternativeto.net/software/maf/
- Wikipedia

Also, marketing both versions at the same time might be a difficult path. I'm unaware of the complexity of merging the two extensions but that would apparently help accessing more people for the SingleFileZ version.

**Comment by gildas-lormeau** [comment]
I am aware that SingleFileZ has a marketing problem. Naively, I thought that the technical innovation would be more or less enough and that it would also provoke discussions. For example, putting binary in HTML may shock some purists (and that's understandable). But it didn't happen like that at all. So I still wonder if SingleFileZ really meets a need compared to SingleFile for most people. For the moment, I consider SingleFileZ as a gift to those who read the SingleFile documentation, because they deserve it :).

Maybe indeed the future of SingleFileZ is to be merged with SingleFile. It's a bit of work but it should be possible. Ideally, I would have liked to finish porting SingleFile to Manifest V3 before thinking about it but I'm still waiting for Google to fix the associated bugs.

**Comment by maliayas** [comment]
> I consider SingleFileZ as a gift to those who read the SingleFile documentation

Thanks, it's really a gift :) I love the innovation there. Being able to unzip the ouput for various purposes (editing etc.) gives great power. That's not the case with SingleFile (yeah, avg user may not care about this)

Also selection saving feature is great tech. Thanks again for publishing these great tools.

**Comment by gildas-lormeau** [comment, marked as answer]
For the record, SingleFileZ has (finally) been merged with SingleFile.

**Reply by maliayas** [reply]
Thanks for this!
